# Bayesian A/B Testing Framework — End-to-End Demonstration

## Objective

This notebook demonstrates the complete Bayesian A/B testing pipeline developed throughout the project.

The framework integrates:

- Frequentist hypothesis testing
- Bayesian posterior inference
- Monte Carlo simulation
- Business decision recommendation

The goal is to analyze an A/B experiment from raw conversion counts to a final business recommendation using a unified statistical framework.

In [18]:
import sys
from pathlib import Path

project_root = Path.cwd().parent
sys.path.append(str(project_root))
from src.ab_framework import ABTestFramework
from src.comparison import comparison_plot

In [19]:
# Experiment Data

visitors_A = 1000
conversions_A = 52

visitors_B = 1000
conversions_B = 61

prior_alpha = 1
prior_beta = 1

In [20]:
framework = ABTestFramework()
results = framework.run_full_analysis(
    n_A=visitors_A,
    conv_A=conversions_A,
    n_B=visitors_B,
    conv_B=conversions_B,
    prior_A=(1,1),
    prior_B=(1,1)
)

In [21]:
frequentist = results["frequentist"]
z = results["frequentist"]["z_statistic"]
p = results["frequentist"]["p_value"]
ci = results["frequentist"]["confidence_interval"]
print("Frequentist Analysis")
print("-" * 40)
print(f"Z-statistic          : {frequentist['z_statistic']:.4f}")
print(f"P-value              : {frequentist['p_value']:.4f}")
print(f"95% Confidence Interval : {frequentist['confidence_interval']}")

Frequentist Analysis
----------------------------------------
Z-statistic          : 0.8716
P-value              : 0.3834
95% Confidence Interval : (np.float64(-0.011233721205341253), np.float64(0.029233721205341255))


In [22]:
bayesian = results["bayesian"]
print("Bayesian Analysis")
print("-"*40)


print(f"Posterior Mean A : {bayesian['A']['mean']:.4f}")
print(f"Posterior Mean B : {bayesian['B']['mean']:.4f}")

print(f"Credible Interval A : {bayesian['A']['credible_interval']}")
print(f"Credible Interval B : {bayesian['B']['credible_interval']}")

Bayesian Analysis
----------------------------------------
Posterior Mean A : 0.0529
Posterior Mean B : 0.0619
Credible Interval A : (np.float64(0.03990909031631707), np.float64(0.06756800569908047))
Credible Interval B : (np.float64(0.04781370094915098), np.float64(0.07759387223541465))


In [23]:
monte_carlo = results["monte_carlo"]
print("Monte Carlo Analysis")
print("-" * 40)

print(f"P(B > A)        : {monte_carlo['probability_B_beats_A']:.4%}")
print(f"Expected Uplift : {monte_carlo['expected_uplift']:.4%}")
print(f"Expected Loss   : {monte_carlo['expected_loss']:.4%}")

Monte Carlo Analysis
----------------------------------------
P(B > A)        : 80.8980%
Expected Uplift : 0.9035%
Expected Loss   : 0.1099%


Variant B shows a higher observed conversion rate than Variant A (6.1% vs. 5.2%), but the evidence is not yet strong enough to recommend deployment. The frequentist analysis does not find a statistically significant difference (p = 0.383), and the Bayesian analysis estimates an 80.9% probability that B is better—moderate, but below common decision thresholds such as 95% or 99%. While the expected uplift is encouraging (0.9%), additional data should be collected before making a production decision.

In [25]:
recommendation = results["recommendation"]
print(recommendation)

{'recommendation': 'Continue Experiment', 'reason': 'Current evidence is inconclusive. Collect more data.'}


Variant B demonstrates a higher observed conversion rate than Variant A (6.1% vs. 5.2%). However, the evidence is not yet strong enough to recommend deployment. The frequentist analysis does not find a statistically significant difference (p = 0.3834), and the Bayesian analysis estimates an 80.9% probability that B is better, which is below typical deployment thresholds such as 95% or 99%. While the expected uplift is encouraging (approximately 0.90 percentage points), additional data should be collected before making a production decision.
